cmd/upgrade/accountroles/accountroles.go

```diff
@@ -0,0 +1,160 @@
+	Run: run,
+}
+
+func init() {
+	flags := Cmd.Flags()
+
+	flags.StringVar(
+		&args.prefix,
+		"prefix",
+		aws.DefaultPrefix,
+		"User-defined prefix for all generated AWS resources",
+	)
+	Cmd.MarkFlagRequired("prefix")
+	confirm.AddFlag(flags)
+	interactive.AddFlag(flags)
+}
+
+func run(cmd *cobra.Command, argv []string) {
+	reporter := rprtr.CreateReporterOrExit()
+	logger := logging.CreateLoggerOrExit(reporter)
+
+	prefix := args.prefix
+
+	// Create the AWS client:
+	awsClient, err := aws.NewClient().
+		Logger(logger).
+		Build()
+	if err != nil {
+		reporter.Errorf("Failed to create AWS client: %v", err)
+		os.Exit(1)
+	}
+
+	// Create the client for the OCM API:
+	ocmClient, err := ocm.NewClient().
+		Logger(logger).
+		Build()
+	if err != nil {
+		reporter.Errorf("Failed to create OCM connection: %v", err)
+		os.Exit(1)
+	}
+	defer func() {
+		err = ocmClient.Close()
+		if err != nil {
+			reporter.Errorf("Failed to close OCM connection: %v", err)
+		}
+	}()
+
+	creator, err := awsClient.GetCreator()
+	if err != nil {
+		reporter.Errorf("Failed to get IAM credentials: %s", err)
+		os.Exit(1)
+	}
+
+	isUpgradeNeed, err := awsClient.IsUpgradedNeededForRole(prefix, creator.AccountID)
+	if err != nil {
+		reporter.Errorf("%s", err)
+		os.Exit(1)
+	}
+	if !isUpgradeNeed {
+		reporter.Infof("Account role with the prefix '%s' is already up-to-date.", prefix)
+		os.Exit(1)
+	}
+	reporter.Infof("Starting to upgrade the policies")
+
+	err = upgradeAccountRolePolicies(reporter, awsClient, prefix, creator.AccountID)
+	if err != nil {
+		reporter.Errorf("%s", err)
+		os.Exit(1)
+	}
+}
+
+func upgradeAccountRolePolicies(reporter *rprtr.Object, awsClient aws.Client, prefix string, accountID string) error {
+	for file, role := range aws.AccountRoles {
+		name := sts.GetRoleName(prefix, role.Name)
+		if !confirm.Prompt(true, "Upgrade the '%s' role polices to version %s?", name,
+			aws.DefaultPolicyVersion) {
+			continue
+		}
+		filename := fmt.Sprintf("sts_%s_permission_policy.json", file)
+		path := fmt.Sprintf("templates/policies/%s", filename)
+		policyARN := sts.GetPolicyARN(accountID, fmt.Sprintf("%s-Policy", name))
+
+		policy, err := aws.ReadPolicyDocument(path)
+		if err != nil {
+			return err
+		}
+		reporter.Debugf("Creating permission policy '%s'", policyARN)
+		policyARN, err = awsClient.EnsurePolicy(policyARN, string(policy),
+			aws.DefaultPolicyVersion, map[string]string{
+				tags.OpenShiftVersion: aws.DefaultPolicyVersion,
+				tags.RolePrefix:       prefix,
+				tags.RoleType:         file,
+			})
+		if err != nil {
+			return err
+		}
+		reporter.Debugf("Attaching permission policy to role '%s'", filename)
+		err = awsClient.AttachRolePolicy(name, policyARN)
+		if err != nil {
+			return err
+		}
+		err = awsClient.UpdateTag(name)
+		if err != nil {
+			return err
+		}
+		//is this approach  ok? or create a policy with a different name and delete the one with the old name?
+		err = awsClient.DeleteOldPolicy(policyARN)
```

Comment:
@vkareh  we are already creating it as new policy version so techinically we are reusing the policy my question for should we delete the old policy version

Reply:
> @vkareh we are already creating it as new policy version so technically we are reusing the policy my question for should we delete the old policy version
```

func (c *awsClient) DeleteOldPolicy(policyARN string) error {
	versions, err := c.iamClient.ListPolicyVersions(&iam.ListPolicyVersionsInput{PolicyArn: aws.String(policyARN)})
	if err != nil {
		return err
	}
	for _, version := range versions.Versions {
		if aws.BoolValue(version.IsDefaultVersion) {
			continue
		}
		_, err := c.iamClient.DeletePolicyVersion(&iam.DeletePolicyVersionInput{
			VersionId: version.VersionId,
			PolicyArn: aws.String(policyARN),
		})
		if err != nil {
			return err
		}
	}
	return nil
}
```


So will the customers have their own versions? currently aws allows up to 5 versions.   is it safe to delete the version alone by reusing the policy without deleting

Reply:
huh, good point. I'd leave it be for now in case users want to audit the difference. By the time we get to 5 policy versions the user can either deal with it manually, or we can provide a better mechanism when that version comes up